coderd/database/pubsub/latency.go

```diff
@@ -0,0 +1,74 @@
+package pubsub
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+)
+
+// LatencyMeasurer is used to measure the send & receive latencies of the underlying Pubsub implementation. We use these
+// measurements to export metrics which can indicate when a Pubsub implementation's queue is overloaded and/or full.
+type LatencyMeasurer struct {
+	// Create unique pubsub channel names so that multiple coderd replicas do not clash when performing latency measurements.
+	channel uuid.UUID
+	logger  slog.Logger
+}
+
+// LatencyMessageLength is the length of a UUIDv4 encoded to hex.
+const LatencyMessageLength = 36
+
+func NewLatencyMeasurer(logger slog.Logger) *LatencyMeasurer {
+	return &LatencyMeasurer{
+		channel: uuid.New(),
+		logger:  logger,
+	}
+}
+
+// Measure takes a given Pubsub implementation, publishes a message & immediately receives it, and returns the observed latency.
+func (lm *LatencyMeasurer) Measure(ctx context.Context, p Pubsub) (send float64, recv float64, err error) {
+	var (
+		start time.Time
+		res   = make(chan float64, 1)
+	)
+
+	msg := []byte(uuid.New().String())
+
+	cancel, err := p.Subscribe(lm.latencyChannelName(), func(ctx context.Context, in []byte) {
+		if !bytes.Equal(in, msg) {
+			lm.logger.Warn(ctx, "received unexpected message", slog.F("got", in), slog.F("expected", msg))
```

Comment:
The initial concern should be obviated now by the background measurement.

I think we could still keep this message around since we should never really see this message. The message itself is meaningless but it'll be useful to differentiate one occurrence from the next. Alternatively we could log each measurement message in `DEBUG` which would allow us to make sense of the actually unexpected one.

Reply:
Yeah, agreed, should be WARN now that it's independent of how many scrapers.  I also like the idea of a DEBUG with the publish, in case we ever have to debug this.